persist/fs/read_test.go

```diff
@@ -407,3 +421,143 @@ func TestReadValidate(t *testing.T) {
 
 	require.NoError(t, r.Close())
 }
+
+func reads(buf *readerDecoderStream, m int) string {
+	var b [1000]byte
+	if int(buf.Remaining()) > len(b) {
+		panic(fmt.Errorf("cannot read all"))
+	}
+
+	nb := 0
+	for {
+		n, err := buf.Read(b[nb : nb+m])
+		nb += n
+		if err == io.EOF {
+			break
+		}
+	}
+	return string(b[0:nb])
+}
+
+func TestDecoderStream(t *testing.T) {
+	var texts [31]string
+	str := ""
+	all := ""
+	for i := 0; i < len(texts)-1; i++ {
+		texts[i] = str + "\n"
+		all += texts[i]
+		str += string(i%26 + 'a')
+	}
+	texts[len(texts)-1] = all
+
+	buf := newReaderDecoderStream()
+	for i := 0; i < len(texts); i++ {
+		text := texts[i]
+		for j := 1; j <= 8; j++ {
+			buf.Reset([]byte(text))
+			s := reads(buf, j)
+			if s != text {
+				t.Errorf("m=%d want=%q got=%q", j, text, s)
+			}
+		}
+	}
+}
+
+func TestDecoderStreamSkip(t *testing.T) {
+	d := []byte{1, 2, 3, 4, 5}
+	expectedDigest := digest.Checksum(d)
+	buf := newReaderDecoderStream()
+	buf.Reset(d)
+	assert.Equal(t, int64(5), buf.Remaining())
+	assert.NoError(t, buf.Skip(3))
+	assert.Equal(t, int64(2), buf.Remaining())
+
+	p := make([]byte, 2)
+	n, err := buf.Read(p)
+	assert.Equal(t, 2, n)
+	assert.NoError(t, err)
+	assert.Equal(t, []byte{4, 5}, p)
+
+	assert.NoError(t, buf.readerWithDigest.Validate(expectedDigest))
+}
+
+func TestDecoderStreamUnreadByte(t *testing.T) {
+	segments := []string{"Hello, ", "world"}
+	got := ""
+	want := strings.Join(segments, "")
+	r := newReaderDecoderStream()
+	r.Reset([]byte(want))
+	// Normal execution.
+	for {
+		b1, err := r.ReadByte()
+		if err != nil {
+			if err != io.EOF {
+				t.Error("unexpected error on ReadByte:", err)
+			}
+			break
+		}
+		got += string(b1)
+		// Put it back and read it again.
+		if err = r.UnreadByte(); err != nil {
+			t.Fatal("unexpected error on UnreadByte:", err)
+		}
+		b2, err := r.ReadByte()
+		if err != nil {
+			t.Fatal("unexpected error reading after unreading:", err)
+		}
+		if b1 != b2 {
+			t.Fatalf("incorrect byte after unread: got %q, want %q", b1, b2)
+		}
+	}
+	if got != want {
+		t.Errorf("got %q, want %q", got, want)
+	}
+}
+
+func TestDecoderStreamUnreadByteMultiple(t *testing.T) {
```

Comment:
+1 for this keeping this test. 

Feels like you're having to test a lot of weird combination of conditions, how would you feel about adding a few property tests for the decoder?

Reply:
I'll open an issue for it, in the interest of expediting this change we'll look at this later.

Reply:
Opened #403